src/main/java/org/alfresco/heartbeat/HBDataCollectorServiceImpl.java

```diff
@@ -63,37 +78,158 @@ public void setHbDataSenderService(HBDataSenderService hbDataSenderService)
         this.hbDataSenderService = hbDataSenderService;
     }
 
+    public void setJobLockService(JobLockService jobLockService)
+    {
+        this.jobLockService = jobLockService;
+    }
+
+    public void setScheduler(Scheduler scheduler)
+    {
+        this.scheduler = scheduler;
+    }
+
+    public void setTestMode(boolean testMode)
+    {
+        this.testMode = testMode;
+    }
+
     /**
      *
-     * Register data collector with this service.
-     * The registered collectors will be called to provide heartbeat data.
+     * Register data collector with this service and start the schedule.
+     * The registered collector will be called to provide heartbeat data.
+     * The collector must be unique with using the collector id.
      *
      * @param collector collector to register
      */
     @Override
     public void registerCollector(HBBaseDataCollector collector)
     {
+        for(HBBaseDataCollector col : collectors)
+        {
+            if(col.getCollectorId().equals(collector.getCollectorId()))
+            {
+                if (logger.isDebugEnabled())
+                {
+                    logger.debug("Didn't register collector because it already exists: " + collector.getCollectorId());
+                }
+                return;
+            }
+        }
         this.collectors.add(collector);
+        try
+        {
+            scheduleCollector(collector);
+        }
+        catch (Exception e)
+        {
+            logger.error("Unable to schedule heart beat", e);
+        }
     }
 
     /**
-     *  Collects and sends data for all registered collectors using the provided sender service.
+     * Listens for license changes.  If a license is change or removed, the heartbeat job is rescheduled.
      */
-    @Override
-    public void collectAndSendData()
+    public synchronized void onLicenseChange(LicenseDescriptor licenseDescriptor)
     {
-        for (HBBaseDataCollector collector : collectors)
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("Update license called");
+        }
+
+        boolean newEnabled = !licenseDescriptor.isHeartBeatDisabled();
+
+        if (newEnabled != this.enabled)
         {
-            List<HBData> data = collector.collectData();
+            if (logger.isDebugEnabled())
+            {
+                logger.debug("State change of heartbeat");
+            }
+            this.enabled = newEnabled;
```

Comment:
This class needs to update the hbDataSenderService enabled state whenever the state changes. 
Could call hbDataSenderService.enable(enabled) wherever this.enabled is set or make a method to do both. 
This method used to do that "public synchronized void enabled(boolean enabled)" We don't need the method to be public any more - this was only used by the HeartBeat class